Initial proposal for secret handling (#2)
package-operator/enhancements: enhancements/E-0001-secret-handling.md

documentation such as release notes or a development roadmap. It should be
possible to collect this information before implementation begins, in order to
avoid requiring implementors to split their attention between writing release
notes and implementing the feature itself. 

A good summary is probably at least a paragraph in length.

Both in this section and below, follow the guidelines of the [documentation
style guide]. In particular, wrap lines to a reasonable length, to make it
easier for reviewers to cite specific portions, and to minimize diff churn on
updates.

[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
-->

## Motivation

<!--
This section is for explicitly listing the motivation, goals, and non-goals of
this Enhancement Proposal. Describe why the change is important and the benefits to users. The
motivation section can optionally provide links to [experience reports] to
demonstrate the interest in a Enhancement Proposal within the wider community.

[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
-->

> - Users deploying a package want to be able to pull package images from private registries, by supplying a reference to an existing secret object containing a valid pull secret.

Reviewer: ```suggestion
- Users deploying a package want to be able to pull package images from private registries.
- Prevent the use of invalid Pull Secrets and provide early and accurate error messages.
```
Try to separate multiple motivations as much as possible, to improve readability.

> - Users crafting a package want to be able to declare secret references, and dynamically reference them with an alias name.

Reviewer: ```suggestion
- Package authors want the ability to declare secrets used within the package.
- Secret references are used to not duplicate sensitive information through PKO APIs.
```

### Non-Goals

<!--
What is out of scope for this enhancement proposal?
Listing non-goals helps to focus discussion and make progress.
-->
- Providing facilities to automagically inject (pull) secret references into managed objects. (For example PodSpecs or (Cluster)Packages.)

## Proposal

<!--
This is where we get down to the specifics of what the proposal actually is.
This should have enough detail that reviewers can understand exactly what
you're proposing, but should not include things like API designs or
implementation. What is the desired outcome and how do we measure success?.
The "Design Details" section below is for the real
nitty-gritty.
-->

- A new object of kind `SecretSync` will be introduced to the in-cluster API of PKO
  - allowing a specified `Secret` source object to be synced to multiple specified namespace/name combinations
  - implementing status reporting so that it can be used as a building block in automations - for example:
    - PKO phases waiting for successful synchronization

>     - (Optional - feasibility unverified) PKO probes reflecting degraded workload health when a synced secret vanishes or during out-of-sync times after a source secret has been changed and the resynchronization hasn't happened yet

Reviewer: Not really understanding this point.
Reacting to status changes is already implemented in PKO itself and works.

Author: What I meant to explore there is:
> The SecretSync status should reflect when source and destination become out of sync.

Sorry for the inprecise wording. :sweat_smile:

  - usage can either happen
    - freely (without using PKO's other API functionalities), maybe solving other problems of users not taken into further consideration in this proposal other than thinking about the performance implications on PKO and the apiserver this might have
    - within packages with an additional guardrail (dubbed `secret-sync-fence`):
      - By default, PKO checks that a `SecretSync` object that is templated by a package in the templating phase references one of the secrets specified in the `(Cluster)Package` object controlling the `(Cluster)ObjectDeployment` about to be created or updated to avoid accidentally syncing the wrong secrets
      - This mechanism should have an opt-out hatch if one has good reasons to include specific `SecretSync` objects with hardcoded sources.

>       - This mechanism will probably turn out to be validatable at packaging- AND run-time. Validating at packaging-time would left-shift surfacing of guardrail violations and save a lot of iteration time!

Reviewer: > This mechanism will probably turn out to be [...]

Don't speculate.
Describe the situation/problem and what steps you want to take to remedy/address.

> Validating at packaging-time would left-shift surfacing of guardrail violations and save a lot of iteration time!

Please describe this for the implementer. Is validation a MUST, where is it OPTIONAL to safe time?

Author: I'll remove this line completely.

      - Important: this guardrail cannot be used for defensive security against malicious packagers because it would be easy to put an evil SecretSync object into a wrapper object like an OpenShift `Template` (or another `Package`), as PKO cannot possibly understand and correctly detect and unpack all possible forms of a "trojan horse" object.

>         - personal recommendation from Josh: If you want to be sure about the authenticity of a package image, implement validation of sigstore signatures on package images.

Reviewer: What does signature verification have to do with this proposal?

Author: I pointed out that this guardrail can only be seen as a convenience helper and not as a real security tool and what direction I would go instead for securing package images.

Reviewer 2: yeah, maybe not falling under this proposal's scope but considering and leveraging sigstore/cosign might be an asset for making sure the resources and their images in the package have not been tampered, for a better suppy-chain security

Author: I'm dropping the guardrail fence completely for the first iteration of the feature.

- The PackageManifest API has to be changed so that `Secret` references can be specified and aliased. References specify:
  - an alias name to be used with newly introduced resolving functions in the package templates
  - when package is installable cluster-scoped: a namespaced name of the source secret
  - when package is installable namespace-scoped: a name of the source secret which is expected in the Package's target namespace

>   - ~if the reference is optional or not (`TODO`: clarify implications - probably better off to start without optional secrets and introduce them later when needed. Big point: how to handle optionality without polling for secret's to appear? The package reconciler doesn't declaratively re-template the ObjectDeployment except, when .spec.image changes)~

Reviewer: I think we never want - on this level - to watch for inputs to be created or changed.
Optional should probably mean that this references is not mandatory to be supplied.

If a secret is referenced, I think that the secret must exist to successfully unpack the image.
If the secret is removed after the package has been unpacked, we _could_ raise a new Condition on the Package to point it out to the user and make debugging easier.

Author: > I think we never want - on this level - to watch for inputs to be created or changed.
> Optional should probably mean that this references is not mandatory to be supplied.

Got it.
Would you argue _for_ including optional references in this proposal?

> If a secret is referenced, I think that the secret must exist to successfully unpack the image.

This is not as easy. The general pattern in kube is "if a secret reference isn't there then the affected controllers will keep retrying until it is there", to not take away forward- and backward-compat by requiring a specific thing to exist before it is used.

I explicitly propose _not_ checking for the existence of referenced secrets at unpack time and let the actual workloads requiring those secrets fail later on, which will be reflected by workload status and pko's probes. This way we minimize stepping on the toes of deployers and their deployment flows.

- The `(Cluster)Package` API has to be changed to
  - allow the specification of an `ImagePullSecret` for pulling the package image itself from a private registry.
  - allow an end-user to supply a list of secret references and bind them to the alias names described in the `PackageManifest` by the packaging person.

- The `kubectl-package` cli has to be changed to
  - include validation for the `secret-sync-fence` guardrail
  - include ways for an end-user to inspect a package image and get a list of secret references that the `(Cluster)Package` requires.
    - This could be part of an overall `kubectl package inspect docker://quay.io/foo/bar-package` command that shows interesting information about a package to the end-user.

> - for handling `ConfigMaps` the described changes could either be duplicated in all affected APIs with `s/Secret/ConfigMap/` or the described API could be made generic (what are the implications of either? I can think of questions about caching and watching/caching too many different object kinds to stay performant).

Reviewer: `ConfigMap` usage should be discussed in it's own concept.

Author: Removed :)

## Design Details

<!--
This section should contain enough information that the specifics of your
change are understandable. This may include API specs (though not always
required) or even code snippets. If there's any ambiguity about HOW your
proposal will be implemented, this is the place to discuss them.
-->

### Changes

#### PackageManifest API changes

The PackageManifest API needs to be changed to include an optional list `.spec.secretRefAliases`, which allows the person writing the object manifests/templates inside the package image to
  - communicate required secret references to the person deploying the package.
  - refer to secrets in their templates using the specified alias.
This effectively decouples package templates and the runtime cluster context and allows the packager to forgo hardcoding magic secret references into the package templates.

```yaml
apiVersion: manifests.package-operator.run/v1alpha1
kind: PackageManifest
metadata:
  name: my-package
spec:
  # Optional: specify a list of aliases for referecing secrets in templates.

>   secretRefAliases:

Reviewer: ```suggestion
  secrets:
```
I would prefer a simpler config key.

Author: done :)

>   - name: tls-keypair-and-certs

Reviewer: I think we should mandate a description to be provided, so users know what this Secret is used for.

Author: I've added an example usage of a `desc` key below. What do you think of it?

TODO: run this by [@thetechnick](https://github.com/thetechnick) again. Should we provide this feature and if we do, should we build an opt out mechanism? I think yes because we don't want to unneccessarily restrict the freedom of what can be in a package.

The package controller should be changed to include the following "synergy" feature for SecretSync objects in Packages.

To prevent unforeseen mishaps, when the rendered output of a package includes a SecretSync Object, PKO must validate that the object reference in `SecretSync.spec.src` points to a secret that is aliased in `.spec.secretRefs`.
This cannot be relied on as a security feature because it is very easy to include wrapper objects (like OpenShift's Template objects) that act as trojan horses, but it serves as a guiding rail that steers package maintainers away from including hard-coded SecretSyncs in their packages and instead honor the user by declaring and aliasing secret references in the PackageManifest, so that they can be supplied dynamically at deployment time.

#### Adapt kubectl-package

Kubectl-package must introduce a new `inspect` command that prints its specified secret requirements.

> The output format is not specified here but it should be open to being expanded with other data about the package and being json/yaml formattable to aid further feature work in the future.

Reviewer: Please create a new concept to outline the inspect command.
While the feature introduced by this proposal makes an inspect command useful, I would not piggy back more features into this concept.

Author: removed

Listing non-goals helps to focus discussion and make progress.
-->
- Providing facilities to automagically inject (pull) secret references into managed objects. (For example PodSpecs or (Cluster)Packages.)

> - Syncing secrets across the MC/HC barried in the HyperShift architecture.

Reviewer 2: Since this was brought up yesterday, and I wanted to cross-post here for posterity: in theory this might work without any extra effort out of the box since _controlplane_ and _dataplane_ can talk to each other via _konnectivity_ proxy?

Author: This would even work without the konnectivity service because  workloads in the MC can directly talk to the HC control-planes hosted in that cluster.

I ruled that functionality out bc of:
- architecture concerns: PKO in the MC does not talk to HC control-planes directly as of today
- scalability concerns: PKO needs to be able to concurrenctly  watch and reconcile the synced secret objects in many hosted control-planes at the same time
- security concerns: Not providing generic capabilities to sync secrets from MC to HC also does not provide the ability to accidentally expose the wrong (platform) secrets to a hosted-cluster
- roadmap concerns: we weren't asked for the capabilities.

Do you think we need this feature?
